File: ios/App/App/Models/ProjectSettingsModels.swift
import Foundation

struct ProjectAction: Codable, Equatable, Hashable, Sendable {
    var id: String
    var name: String
    var command: String
    var icon: String?
}

struct ProjectScripts: Codable, Equatable, Hashable, Sendable {
    var setupScript: String?
    var cleanupScript: String?
    var worktreeCopyPatterns: [String]?
    var actions: [ProjectAction]

    init(
        setupScript: String? = nil,
        cleanupScript: String? = nil,
        worktreeCopyPatterns: [String]? = nil,
        actions: [ProjectAction] = []
    ) {
        self.setupScript = setupScript
        self.cleanupScript = cleanupScript
        self.worktreeCopyPatterns = worktreeCopyPatterns
        self.actions = actions
    }

    private enum CodingKeys: String, CodingKey {
        case setupScript
        case cleanupScript
        case worktreeCopyPatterns
        case actions
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        setupScript = try values.decodeIfPresent(String.self, forKey: .setupScript)
        cleanupScript = try values.decodeIfPresent(String.self, forKey: .cleanupScript)
        worktreeCopyPatterns = try values.decodeIfPresent(
            [String].self, forKey: .worktreeCopyPatterns)
        actions = try values.decodeIfPresent([ProjectAction].self, forKey: .actions) ?? []
    }
}

struct ProjectSearchSettings: Codable, Equatable, Hashable, Sendable {
    var useIgnoreFiles: Bool?
    var exclude: [String: Bool]?
}

struct ProjectWorktreeLocation: Codable, Equatable, Hashable, Sendable {
    enum Mode: String, Codable, Sendable {
        case global
        case projectRelative = "project-relative"
    }

    var mode: Mode?
    var basePath: String?
}

/// A wire dictionary whose ordinary string and debug representations redact values.
struct SensitiveStringMap: Codable, Equatable, Hashable, Sendable,
    CustomStringConvertible, CustomDebugStringConvertible
{
    private let storage: [String: String]

    init(_ storage: [String: String] = [:]) {
        self.storage = storage
    }

    var count: Int { storage.count }
    var keys: [String] { storage.keys.sorted() }

    func secretValue(forKey key: String) -> String? {
        storage[key]
    }

    var description: String {
        "SensitiveStringMap(keys: \(keys), values: <redacted>)"
    }

    var debugDescription: String { description }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        storage = try container.decode([String: String].self)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(storage)
    }
}

enum ProjectMCPTransport: Codable, Equatable, Hashable, Sendable,
    CustomStringConvertible, CustomDebugStringConvertible
{
    case stdio(command: String, args: [String], env: SensitiveStringMap, cwd: String?)
    case http(url: String, headers: SensitiveStringMap)
    case sse(url: String, headers: SensitiveStringMap)

    private enum CodingKeys: String, CodingKey {
        case type
        case command
        case args
        case env
        case cwd
        case url
        case headers
    }

    private enum Kind: String, Codable {
        case stdio
        case http
        case sse
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        switch try values.decode(Kind.self, forKey: .type) {
        case .stdio:
            self = .stdio(
                command: try values.decode(String.self, forKey: .command),
                args: try values.decodeIfPresent([String].self, forKey: .args) ?? [],
                env: try values.decodeIfPresent(SensitiveStringMap.self, forKey: .env) ?? .init(),
                cwd: try values.decodeIfPresent(String.self, forKey: .cwd)
            )
        case .http:
            self = .http(
                url: try values.decode(String.self, forKey: .url),
                headers: try values.decodeIfPresent(SensitiveStringMap.self, forKey: .headers)
                    ?? .init()
            )
        case .sse:
            self = .sse(
                url: try values.decode(String.self, forKey: .url),
                headers: try values.decodeIfPresent(SensitiveStringMap.self, forKey: .headers)
                    ?? .init()
            )
        }
    }

    func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .stdio(let command, let args, let env, let cwd):
            try values.encode(Kind.stdio, forKey: .type)
            try values.encode(command, forKey: .command)
            try values.encode(args, forKey: .args)
            try values.encode(env, forKey: .env)
            try values.encodeIfPresent(cwd, forKey: .cwd)
        case .http(let url, let headers):
            try values.encode(Kind.http, forKey: .type)
            try values.encode(url, forKey: .url)
            try values.encode(headers, forKey: .headers)
        case .sse(let url, let headers):
            try values.encode(Kind.sse, forKey: .type)
            try values.encode(url, forKey: .url)
            try values.encode(headers, forKey: .headers)
        }
    }

    var description: String {
        switch self {
        case .stdio(let command, let args, _, let cwd):
            "ProjectMCPTransport.stdio(command: \(command), args: \(args), env: <redacted>, cwd: \(cwd ?? "nil"))"
        case .http(let url, _):
            "ProjectMCPTransport.http(url: \(url), headers: <redacted>)"
        case .sse(let url, _):
            "ProjectMCPTransport.sse(url: \(url), headers: <redacted>)"
        }
    }

    var debugDescription: String { description }
}

struct ProjectMCPServer: Codable, Equatable, Hashable, Sendable,
    CustomStringConvertible, CustomDebugStringConvertible
{
    var id: String
    var name: String
    var descriptionText: String
    var enabled: Bool
    var timeoutMs: Int
    var disabledTools: [String]?
    var transport: ProjectMCPTransport

    init(
        id: String,
        name: String,
        descriptionText: String = "",
        enabled: Bool = true,
        timeoutMs: Int = 30_000,
        disabledTools: [String]? = nil,
        transport: ProjectMCPTransport
    ) {
        self.id = id
        self.name = name
        self.descriptionText = descriptionText
        self.enabled = enabled
        self.timeoutMs = timeoutMs
        self.disabledTools = disabledTools
        self.transport = transport
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case name
        case descriptionText = "description"
        case enabled
        case timeoutMs
        case disabledTools
        case transport
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        id = try values.decode(String.self, forKey: .id)
        name = try values.decode(String.self, forKey: .name)
        descriptionText = try values.decodeIfPresent(String.self, forKey: .descriptionText) ?? ""
        enabled = try values.decodeIfPresent(Bool.self, forKey: .enabled) ?? true
        timeoutMs = try values.decodeIfPresent(Int.self, forKey: .timeoutMs) ?? 30_000
        disabledTools = try values.decodeIfPresent([String].self, forKey: .disabledTools)
        transport = try values.decode(ProjectMCPTransport.self, forKey: .transport)
    }

    var description: String {
        "ProjectMCPServer(id: \(id), name: \(name), transport: <redacted>)"
    }

    var debugDescription: String { description }
}

struct ProjectSettings: Codable, Equatable, Sendable {
    var mcpServers: [ProjectMCPServer]?
}

File: android/app/src/main/kotlin/com/poracode/app/model/ProjectSettingsModels.kt
package com.poracode.app.model

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonClassDiscriminator

@Serializable
data class ProjectDraftConfig(
    val agentKind: String,
    val model: String = "",
    val effort: String? = null,
    val contextSize: String? = null,
    val fast: Boolean? = null,
    val thinking: Boolean? = null,
    val mode: String? = null,
    val approvalPolicy: String? = null,
    val approvalsReviewer: String? = null,
    val sandboxMode: String? = null,
    val browserMcp: Boolean? = null,
    val crossagentMcp: Boolean? = null,
    val computerUse: Boolean? = null,
    val chromeMcp: Boolean? = null,
    val worktreeMode: Boolean? = null,
)

@Serializable
data class ProjectAction(
    val id: String,
    val name: String,
    val command: String,
    val icon: String? = null,
) {
    override fun toString(): String =
        "ProjectAction(id=$id, name=$name, command=<redacted>, icon=$icon)"
}

@Serializable
data class ProjectScripts(
    val setupScript: String? = null,
    val cleanupScript: String? = null,
    val worktreeCopyPatterns: List<String>? = null,
    val actions: List<ProjectAction> = emptyList(),
) {
    override fun toString(): String =
        "ProjectScripts(setup=${setupScript != null}, cleanup=${cleanupScript != null}, " +
            "copyPatterns=${worktreeCopyPatterns?.size ?: 0}, actions=${actions.size})"
}

@Serializable
data class ProjectSearchSettings(
    val useIgnoreFiles: Boolean? = null,
    val exclude: Map<String, Boolean>? = null,
)

@Serializable
enum class WorktreeStorageMode {
    @SerialName("global")
    GLOBAL,

    @SerialName("project-relative")
    PROJECT_RELATIVE,
}

@Serializable
data class ProjectWorktreeLocation(
    val mode: WorktreeStorageMode? = null,
    val basePath: String? = null,
)

/** Map values that must never appear in generated log/toString output. */
@Serializable(with = SensitiveStringMapSerializer::class)
class SensitiveStringMap internal constructor(
    internal val wireValues: Map<String, String>,
) {
    val size: Int get() = wireValues.size
    val keys: Set<String> get() = wireValues.keys

    fun valueFor(key: String): String? = wireValues[key]

    override fun equals(other: Any?): Boolean =
        other is SensitiveStringMap && wireValues == other.wireValues

    override fun hashCode(): Int = wireValues.hashCode()

    override fun toString(): String = "<redacted:${wireValues.size}>"

    companion object {
        fun empty(): SensitiveStringMap = SensitiveStringMap(emptyMap())
        fun of(values: Map<String, String>): SensitiveStringMap = SensitiveStringMap(values.toMap())
    }
}

object SensitiveStringMapSerializer : KSerializer<SensitiveStringMap> {
    private val delegate = MapSerializer(String.serializer(), String.serializer())
    override val descriptor = delegate.descriptor

    override fun deserialize(decoder: Decoder): SensitiveStringMap =
        SensitiveStringMap(delegate.deserialize(decoder))

    override fun serialize(encoder: Encoder, value: SensitiveStringMap) {
        delegate.serialize(encoder, value.wireValues)
    }
}

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("type")
sealed interface McpTransport

@Serializable
@SerialName("stdio")
data class McpStdioTransport(
    val command: String,
    val args: List<String> = emptyList(),
    val env: SensitiveStringMap = SensitiveStringMap.empty(),
    val cwd: String? = null,
) : McpTransport {
    override fun toString(): String = "McpStdioTransport(<redacted>)"
}

@Serializable
@SerialName("http")
data class McpHttpTransport(
    val url: String,
    val headers: SensitiveStringMap = SensitiveStringMap.empty(),
) : McpTransport {
    override fun toString(): String = "McpHttpTransport(<redacted>)"
}

@Serializable
@SerialName("sse")
data class McpSseTransport(
    val url: String,
    val headers: SensitiveStringMap = SensitiveStringMap.empty(),
) : McpTransport {
    override fun toString(): String = "McpSseTransport(<redacted>)"
}

@Serializable
data class McpServer(
    val id: String,
    val name: String,
    val description: String = "",
    val enabled: Boolean = true,
    val timeoutMs: Int = DEFAULT_TIMEOUT_MS,
    val disabledTools: List<String>? = null,
    val transport: McpTransport,
) {
    override fun toString(): String =
        "McpServer(id=$id, name=$name, enabled=$enabled, transport=${transport::class.simpleName}, " +
            "sensitiveConfig=<redacted>)"

    companion object {
        const val DEFAULT_TIMEOUT_MS = 30_000
    }
}

/** Kept separate from RemoteProject because MCP configuration may contain secrets. */
@Serializable
data class ProjectSettings(
    val mcpServers: List<McpServer>? = null,
)
